### PR TITLE
add accelerate example

### PR DESCRIPTION
**Pull request by edbeeching**
## Types of changes

- [ ] Bug fix
- [x] New feature
- [ ] New algorithm
- [ ] Documentation

## Checklist:

- [ ] I've read the [CONTRIBUTION](https://docs.cleanrl.dev/contribution/) guide (**required**).
- [ ] I have ensured `pre-commit run --all-files` passes (**required**).
- [ ] I have updated the tests accordingly (if applicable).
- [ ] I have updated the documentation and previewed the changes via `mkdocs serve`.
    - [ ] I have explained note-worthy implementation details.
    - [ ] I have explained the logged metrics.
    - [ ] I have added links to the original paper and related papers.

If you need to run benchmark experiments for a performance-impacting changes:

- [ ] I have contacted @vwxyzjn to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark).
- [ ] I have used the [benchmark utility](/get-started/benchmark-utility/) to submit the tracked experiments to the [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) W&B project, optionally with `--capture_video`.
- [ ] I have performed RLops with `python -m openrlbenchmark.rlops`.
    - For new feature or bug fix:
        - [ ] I have used the RLops utility to understand the performance impact of the changes and confirmed there is no regression.
    - For new algorithm:
        - [ ] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [ ] I have added the learning curves generated by the `python -m openrlbenchmark.rlops` utility to the documentation.
    - [ ] I have added links to the tracked experiments in W&B, generated by `python -m openrlbenchmark.rlops ....your_args... --report`,  to the documentation.

### CODE REVIEW

File: cleanrl/ppo_atari_accelerate.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+
+    # Algorithm specific arguments
+    env_id: str = "BreakoutNoFrameskip-v4"
+    """the id of the environment"""
+    total_timesteps: int = 10000000
+    """total timesteps of the experiments"""
+    learning_rate: float = 2.5e-4
+    """the learning rate of the optimizer"""
+    num_envs: int = 8
+    """the number of parallel game environments"""
```
Comment (vwxyzjn) [inline cleanrl/ppo_atari_accelerate.py:54]:
```suggestion
    local_num_envs: int = 8
    """the number of parallel game environments (in the local rank)"""
```

File: cleanrl/ppo_atari_accelerate.py [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+        writer = SummaryWriter(f"runs/{run_name}")
+        writer.add_text(
+            "hyperparameters",
+            "|param|value|\n|-|-|\n%s" % ("\n".join([f"|{key}|{value}|" for key, value in vars(args).items()])),
+        )
+
+    # TRY NOT TO MODIFY: seeding
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed)
```
Comment (vwxyzjn) [inline cleanrl/ppo_atari_accelerate.py:181]: This is tricky: you'd want to use a different seed so that the environments steps are different.
```suggestion
    # CRUCIAL: note that we needed to pass a different seed for each data parallelism worker
    args.seed += accelerator.process_index * 100003  # Prime
    random.seed(args.seed)
    np.random.seed(args.seed)
    torch.manual_seed(args.seed - local_rank)
```

File: cleanrl/ppo_atari_accelerate.py [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    torch.backends.cudnn.deterministic = args.torch_deterministic
+
+    # env setup
+    envs = gym.vector.SyncVectorEnv(
+        [make_env(args.env_id, i, args.capture_video, run_name) for i in range(args.num_envs)],
```
Comment (vwxyzjn) [inline cleanrl/ppo_atari_accelerate.py:186]:
```suggestion
        [make_env(args.env_id, i, args.capture_video, run_name) for i in range(args.local_num_envs)],
```

File: cleanrl/ppo_atari_accelerate.py [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+    )
+    assert isinstance(envs.single_action_space, gym.spaces.Discrete), "only discrete action space is supported"
+
+    agent = Agent(envs)
+    agent = accelerator.prepare(agent)
+    
+    device = accelerator.device
+    optimizer = optim.Adam(agent.parameters(), lr=args.learning_rate, eps=1e-5)
```
Comment (vwxyzjn) [inline cleanrl/ppo_atari_accelerate.py:194]:
```suggestion
    agent = Agent(envs).to(device)
    torch.manual_seed(args.seed)
    optimizer = optim.Adam(agent.parameters(), lr=args.learning_rate, eps=1e-5)
    agent, optimizer = accelerator.prepare(agent, optimizer)
    device = accelerator.device
```

File: cleanrl/ppo_atari_accelerate.py [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+
+    # ALGO Logic: Storage setup
+    obs = torch.zeros((args.num_steps, args.num_envs) + envs.single_observation_space.shape).to(device)
+    actions = torch.zeros((args.num_steps, args.num_envs) + envs.single_action_space.shape).to(device)
+    logprobs = torch.zeros((args.num_steps, args.num_envs)).to(device)
+    rewards = torch.zeros((args.num_steps, args.num_envs)).to(device)
+    dones = torch.zeros((args.num_steps, args.num_envs)).to(device)
+    values = torch.zeros((args.num_steps, args.num_envs)).to(device)
```
Comment (vwxyzjn) [inline cleanrl/ppo_atari_accelerate.py:202]:
```suggestion
    obs = torch.zeros((args.num_steps, args.local_num_envs) + envs.single_observation_space.shape).to(device)
    actions = torch.zeros((args.num_steps, args.local_num_envs) + envs.single_action_space.shape).to(device)
    logprobs = torch.zeros((args.num_steps, args.local_num_envs)).to(device)
    rewards = torch.zeros((args.num_steps, args.local_num_envs)).to(device)
    dones = torch.zeros((args.num_steps, args.local_num_envs)).to(device)
    values = torch.zeros((args.num_steps, args.local_num_envs)).to(device)
```

File: cleanrl/ppo_atari_accelerate.py [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+
+    # TRY NOT TO MODIFY: start the game
+    global_step = 0
+    start_time = time.time()
+    next_obs, _ = envs.reset(seed=args.seed)
+    next_obs = torch.Tensor(next_obs).to(device)
+    next_done = torch.zeros(args.num_envs).to(device)
```
Comment (vwxyzjn) [inline cleanrl/ppo_atari_accelerate.py:209]:
```suggestion
    next_done = torch.zeros(args.local_num_envs).to(device)
```

File: cleanrl/ppo_atari_accelerate.py [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+            next_obs, next_done = torch.Tensor(next_obs).to(device), torch.Tensor(next_done).to(device)
+
+            
+            if "final_info" in infos and accelerator.is_main_process:
+                for info in infos["final_info"]:
+                    if info and "episode" in info:
+                        print(f"global_step={global_step}, episodic_return={info['episode']['r']}")
+                        writer.add_scalar("charts/episodic_return", info["episode"]["r"], global_step)
+                        writer.add_scalar("charts/episodic_length", info["episode"]["l"], global_step)
+
```
Comment (vwxyzjn) [inline cleanrl/ppo_atari_accelerate.py:243]:
```suggestion
            if not writer:
                continue

            if "final_info" in infos:
                for info in infos["final_info"]:
                    if info and "episode" in info:
                        print(f"global_step={global_step}, episodic_return={info['episode']['r']}")
                        writer.add_scalar("charts/episodic_return", info["episode"]["r"], global_step)
                        writer.add_scalar("charts/episodic_length", info["episode"]["l"], global_step)

```

File: cleanrl/ppo_atari_accelerate.py [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+        # bootstrap value if not done
```
Comment (vwxyzjn) [inline cleanrl/ppo_atari_accelerate.py:244]:
```suggestion
        print(
            f"local_rank: {local_rank}, action.sum(): {action.sum()}, iteration: {iteration}, agent.actor.weight.sum(): {agent.actor.weight.sum()}"
        )
        # bootstrap value if not done
```

File: cleanrl/ppo_atari_accelerate.py [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+        b_returns = returns.reshape(-1)
+        b_values = values.reshape(-1)
+
+        # Optimizing the policy and value network
+        b_inds = np.arange(args.batch_size)
+        clipfracs = []
+        for epoch in range(args.update_epochs):
+            np.random.shuffle(b_inds)
+            for start in range(0, args.batch_size, args.minibatch_size):
+                end = start + args.minibatch_size
```
Comment (vwxyzjn) [inline cleanrl/ppo_atari_accelerate.py:274]:
```suggestion
        b_inds = np.arange(args.local_batch_size)
        clipfracs = []
        for epoch in range(args.update_epochs):
            np.random.shuffle(b_inds)
            for start in range(0, args.local_batch_size, args.local_minibatch_size):
                end = start + args.local_minibatch_size
                mb_inds = b_inds[start:end]
```

File: cleanrl/ppo_atari_accelerate.py [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+                    v_loss = 0.5 * v_loss_max.mean()
+                else:
+                    v_loss = 0.5 * ((newvalue - b_returns[mb_inds]) ** 2).mean()
+
+                entropy_loss = entropy.mean()
+                loss = pg_loss - args.ent_coef * entropy_loss + v_loss * args.vf_coef
+
+                optimizer.zero_grad()
+                accelerator.backward(loss)
+                nn.utils.clip_grad_norm_(agent.parameters(), args.max_grad_norm)
```
Comment (vwxyzjn) [inline cleanrl/ppo_atari_accelerate.py:316]:
```suggestion
                accelerator.clip_grad_norm_(agent.parameters(), args.max_grad_norm)
```

File: cleanrl/ppo_atari_accelerate.py [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+            writer.add_scalar("losses/approx_kl", approx_kl.item(), global_step)
+            writer.add_scalar("losses/clipfrac", np.mean(clipfracs), global_step)
+            writer.add_scalar("losses/explained_variance", explained_var, global_step)
+            print("SPS:", int(global_step / (time.time() - start_time)))
+            writer.add_scalar("charts/SPS", int(global_step / (time.time() - start_time)), global_step)
+
+    envs.close()
+    
+    if accelerator.is_main_process:
+        writer.close()
```
Comment (vwxyzjn) [inline cleanrl/ppo_atari_accelerate.py:342]:
```suggestion
        writer.close()
        if args.track:
            wandb.finish()
```

**Review by vwxyzjn — COMMENTED** [review]
Thanks for the PR! I had something quick whip up here https://gist.github.com/vwxyzjn/62197ac158bfa648e3251994bf8ccd3e and made a comparison and added some comments :)
